File: Python-Solution/03-ArraysStringsHashTables/IsUnique.py
"""
 * Dado un método que recibe una String, comprobar si todos los caracteres son únicos o no.
 *
 * isUnique("abcde") => true;
 * isUnique("abcded") => false;

 Solucion a utilizar: HashTable
 Complejidad -> O(N) - Lineal
 --
 Se puede mejorar a O(1) - Constante, si tomamos como referencia los simbolos del ASCII, 
 Si el largo del string es mayor al numero de ASCII, entonces si tenemos valores repetidos.
 Sino, siempre recorreremos los mismos numeros sin incrementar la complejidad.
 --
"""

def IsUnique_HashTable(string) -> bool:
    # creamos la tabla hash
    hashtable = {}
    for i,e in enumerate(string):
        if e in hashtable.keys():
            return False
        hashtable[e] = i
    return True

def IsUnique_MejorComplejidad(string) -> bool:
    NUMBER_OF_CHARS = 128 # ASCII

    if len(string) > NUMBER_OF_CHARS:
        return False

    hashtable = {}
    for i,e in enumerate(string):
        if e in hashtable.keys():
            return False
        hashtable[e] = i

    return True

print(IsUnique_MejorComplejidad("abcde"))
print(IsUnique_MejorComplejidad("abcdefghijklmnopqrstuvwxyzo"))

File: 03-ArraysStringsHashTables/src/main/java/com/danielblanco/algoritmosestructuras/arraysstringshashtables/_01_is_unique/IsUnique.java
package com.danielblanco.algoritmosestructuras.arraysstringshashtables._01_is_unique;

import java.util.HashSet;
import java.util.Set;

/*
 * Dado un método que recibe una String, comprobar si todos los caracteres son únicos o no.
 *
 * isUnique("abcde") => true;
 * isUnique("abcded") => false;
 */
public class IsUnique {

  private static final int NUMBER_OF_CHAR = 128; // ASCII

  public boolean isUnique(String s) {
    if (s.length() > NUMBER_OF_CHAR)
      return false;

    Set<Character> chars = new HashSet<>();
    for (Character c : s.toCharArray()) {
      if (chars.contains(c))
        return false;
      chars.add(c);
    }

    return true;
  }
}
